### PR TITLE
Add script to check annotations.csv

### PR DESCRIPTION
**Pull request by dominikl**
Checks if all images of an IDR project have an entry in an annotations.csv file. Images which don't, will be printed to stdout.

### CODE REVIEW

**Comment by pwalczysko** [comment]
> the two scripts missing_image_annotations.py and missing_well_annotations.py

What two scripts please ?

**Comment by dominikl** [comment]
The two scripts in https://github.com/IDR/idr-utils/tree/master/scripts/annotate . They're very similar, but are doing not quite the same thing. However could be replaced by the script of this PR with an additional option in future.

**Comment by dominikl** [comment]
Improved script so that it can now check images against an annotation.csv file or simply check if all images have annotations attached (if the annotation.csv file parameter is omitted). This makes the two scripts missing_image_annotations.py and missing_well_annotations.py superfluous, hence this PR removes them.

**Comment by pwalczysko** [comment]
```
python check_annotations.py -h
Traceback (most recent call last):
  File "check_annotations.py", line 83, in <module>
    if projectId:
NameError: name 'projectId' is not defined
```

As discussed, please make the ``-h`` to deliver a helpful output.

**Comment by pwalczysko** [comment]
Problem:

In the idr0082, which consists of 3 Datasets, each dataset has 112 images, and none of the images have annotations. 
But in the output below, I am getting the listing of only 224 images, from the 2nd and 3rd dataset. The first dataset is missing.

``(venv3) [omero-server@prod87-omeroreadwrite tmp]$ python check_annotations.py Project:1251``

<details>

```
Missing annotations (Dataset, Image, ID):
Lung Carcinoma Overview,S11_HandE.ndpi [macro image],9846358
Lung Carcinoma Overview,S12_HandE.ndpi [macro image],9846511
Lung Carcinoma Overview,S133_HandE.ndpi [macro image],9846499
Lung Carcinoma Overview,S134_HandE.ndpi [macro image],9846265
Lung Carcinoma Overview,S135_HandE.ndpi [macro image],9846220
Lung Carcinoma Overview,S137_HandE.ndpi [macro image],9846484
Lung Carcinoma Overview,S138_HandE.ndpi [macro image],9846424
Lung Carcinoma Overview,S13_HandE.ndpi [macro image],9846292
Lung Carcinoma Overview,S140_HandE.ndpi [macro image],9846466
Lung Carcinoma Overview,S142_HandE.ndpi [macro image],9846250
Lung Carcinoma Overview,S143_HandE.ndpi [macro image],9846430
Lung Carcinoma Overview,S144_HandE.ndpi [macro image],9846415
Lung Carcinoma Overview,S145_HandE.ndpi [macro image],9846352
Lung Carcinoma Overview,S146_HandE.ndpi [macro image],9846337
Lung Carcinoma Overview,S147_HandE.ndpi [macro image],9846355
Lung Carcinoma Overview,S148_HandE.ndpi [macro image],9846289
Lung Carcinoma Overview,S149_HandE.ndpi [macro image],9846370
Lung Carcinoma Overview,S14_HandE.ndpi [macro image],9846229
Lung Carcinoma Overview,S150_HandE.ndpi [macro image],9846460
Lung Carcinoma Overview,S151_HandE.ndpi [macro image],9846529
Lung Carcinoma Overview,S152_HandE.ndpi [macro image],9846301
Lung Carcinoma Overview,S153_HandE.ndpi [macro image],9846394
Lung Carcinoma Overview,S154_HandE.ndpi [macro image],9846532
Lung Carcinoma Overview,S156_HandE.ndpi [macro image],9846310
Lung Carcinoma Overview,S157_HandE.ndpi [macro image],9846469
Lung Carcinoma Overview,S158_HandE.ndpi [macro image],9846445
Lung Carcinoma Overview,S159_HandE.ndpi [macro image],9846490
Lung Carcinoma Overview,S160_HandE.ndpi [macro image],9846502
Lung Carcinoma Overview,S161_HandE.ndpi [macro image],9846271
Lung Carcinoma Overview,S162_HandE.ndpi [macro image],9846202
Lung Carcinoma Overview,S163_HandE.ndpi [macro image],9846448
Lung Carcinoma Overview,S164_HandE.ndpi [macro image],9846391
Lung Carcinoma Overview,S165_HandE.ndpi [macro image],9846376
Lung Carcinoma Overview,S166_HandE.ndpi [macro image],9846325
Lung Carcinoma Overview,S167_HandE.ndpi [macro image],9846226
Lung Carcinoma Overview,S168_HandE.ndpi [macro image],9846367
Lung Carcinoma Overview,S169_HandE.ndpi [macro image],9846247
Lung Carcinoma Overview,S16_HandE.ndpi [macro image],9846277
Lung Carcinoma Overview,S170_HandE.ndpi [macro image],9846217
Lung Carcinoma Overview,S17_HandE.ndpi [macro image],9846373
Lung Carcinoma Overview,S18_HandE.ndpi [macro image],9846493
Lung Carcinoma Overview,S19_HandE.ndpi [macro image],9846349
Lung Carcinoma Overview,S1_HandE.ndpi [macro image],9846304
Lung Carcinoma Overview,S20_HandE.ndpi [macro image],9846478
Lung Carcinoma Overview,S21_HandE.ndpi [macro image],9846280
Lung Carcinoma Overview,S22_HandE.ndpi [macro image],9846454
Lung Carcinoma Overview,S23_HandE.ndpi [macro image],9846232
Lung Carcinoma Overview,S24_HandE.ndpi [macro image],9846298
Lung Carcinoma Overview,S25_HandE.ndpi [macro image],9846409
Lung Carcinoma Overview,S26_HandE.ndpi [macro image],9846382
Lung Carcinoma Overview,S27_HandE.ndpi [macro image],9846343
Lung Carcinoma Overview,S28_HandE.ndpi [macro image],9846436
Lung Carcinoma Overview,S29_HandE.ndpi [macro image],9846208
Lung Carcinoma Overview,S2_HandE.ndpi [macro image],9846388
Lung Carcinoma Overview,S32_HandE.ndpi [macro image],9846481
Lung Carcinoma Overview,S36_HandE.ndpi [macro image],9846241
Lung Carcinoma Overview,S38_HandE.ndpi [macro image],9846340
Lung Carcinoma Overview,S39_HandE.ndpi [macro image],9846319
Lung Carcinoma Overview,S3_HandE.ndpi [macro image],9846238
Lung Carcinoma Overview,S40_HandE.ndpi [macro image],9846253
Lung Carcinoma Overview,S41_HandE.ndpi [macro image],9846346
Lung Carcinoma Overview,S42_HandE.ndpi [macro image],9846496
Lung Carcinoma Overview,S43_HandE.ndpi [macro image],9846421
Lung Carcinoma Overview,S46_HandE.ndpi [macro image],9846427
Lung Carcinoma Overview,S47_HandE.ndpi [macro image],9846256
Lung Carcinoma Overview,S4_HandE.ndpi [macro image],9846487
Lung Carcinoma Overview,S50_HandE.ndpi [macro image],9846307
Lung Carcinoma Overview,S51_HandE.ndpi [macro image],9846328
Lung Carcinoma Overview,S53_HandE.ndpi [macro image],9846526
Lung Carcinoma Overview,S55_HandE.ndpi [macro image],9846463
Lung Carcinoma Overview,S56_HandE.ndpi [macro image],9846313
Lung Carcinoma Overview,S57_HandE.ndpi [macro image],9846334
Lung Carcinoma Overview,S58_HandE.ndpi [macro image],9846418
Lung Carcinoma Overview,S59_HandE.ndpi [macro image],9846322
Lung Carcinoma Overview,S60_HandE.ndpi [macro image],9846520
Lung Carcinoma Overview,S61_HandE.ndpi [macro image],9846364
Lung Carcinoma Overview,S62_HandE.ndpi [macro image],9846523
Lung Carcinoma Overview,S63_HandE.ndpi [macro image],9846223
Lung Carcinoma Overview,S65_HandE.ndpi [macro image],9846508
Lung Carcinoma Overview,S66_HandE.ndpi [macro image],9846475
Lung Carcinoma Overview,S67_HandE.ndpi [macro image],9846472
Lung Carcinoma Overview,S68_HandE.ndpi [macro image],9846403
Lung Carcinoma Overview,S69_HandE.ndpi [macro image],9846442
Lung Carcinoma Overview,S6_HandE.ndpi [macro image],9846268
Lung Carcinoma Overview,S71_HandE.ndpi [macro image],9846397
Lung Carcinoma Overview,S72_HandE.ndpi [macro image],9846505
Lung Carcinoma Overview,S73_HandE.ndpi [macro image],9846316
Lung Carcinoma Overview,S74_HandE.ndpi [macro image],9846400
Lung Carcinoma Overview,S76_HandE.ndpi [macro image],9846259
Lung Carcinoma Overview,S77_HandE.ndpi [macro image],9846451
Lung Carcinoma Overview,S78_HandE.ndpi [macro image],9846439
Lung Carcinoma Overview,S79_HandE.ndpi [macro image],9846235
Lung Carcinoma Overview,S7_HandE.ndpi [macro image],9846385
Lung Carcinoma Overview,S80_HandE.ndpi [macro image],9846379
Lung Carcinoma Overview,S81_HandE.ndpi [macro image],9846205
Lung Carcinoma Overview,S82_HandE.ndpi [macro image],9846517
Lung Carcinoma Overview,S83_HandE.ndpi [macro image],9846286
Lung Carcinoma Overview,S84_HandE.ndpi [macro image],9846244
Lung Carcinoma Overview,S85_HandE.ndpi [macro image],9846433
Lung Carcinoma Overview,S86_HandE.ndpi [macro image],9846361
Lung Carcinoma Overview,S87_HandE.ndpi [macro image],9846514
Lung Carcinoma Overview,S88_HandE.ndpi [macro image],9846535
Lung Carcinoma Overview,S89_HandE.ndpi [macro image],9846214
Lung Carcinoma Overview,S8_HandE.ndpi [macro image],9846262
Lung Carcinoma Overview,S90_HandE.ndpi [macro image],9846457
Lung Carcinoma Overview,S91_HandE.ndpi [macro image],9846283
Lung Carcinoma Overview,S92_HandE.ndpi [macro image],9846412
Lung Carcinoma Overview,S93_HandE.ndpi [macro image],9846274
Lung Carcinoma Overview,S94_HandE.ndpi [macro image],9846211
Lung Carcinoma Overview,S96_HandE.ndpi [macro image],9846406
Lung Carcinoma Overview,S97_HandE.ndpi [macro image],9846331
Lung Carcinoma Overview,S9_HandE.ndpi [macro image],9846295
Lung Carcinoma Overview Mask,S11_HandE.ndpi [macro mask image],9846359
Lung Carcinoma Overview Mask,S12_HandE.ndpi [macro mask image],9846512
Lung Carcinoma Overview Mask,S133_HandE.ndpi [macro mask image],9846500
Lung Carcinoma Overview Mask,S134_HandE.ndpi [macro mask image],9846266
Lung Carcinoma Overview Mask,S135_HandE.ndpi [macro mask image],9846221
Lung Carcinoma Overview Mask,S137_HandE.ndpi [macro mask image],9846485
Lung Carcinoma Overview Mask,S138_HandE.ndpi [macro mask image],9846425
Lung Carcinoma Overview Mask,S13_HandE.ndpi [macro mask image],9846293
Lung Carcinoma Overview Mask,S140_HandE.ndpi [macro mask image],9846467
Lung Carcinoma Overview Mask,S142_HandE.ndpi [macro mask image],9846251
Lung Carcinoma Overview Mask,S143_HandE.ndpi [macro mask image],9846431
Lung Carcinoma Overview Mask,S144_HandE.ndpi [macro mask image],9846416
Lung Carcinoma Overview Mask,S145_HandE.ndpi [macro mask image],9846353
Lung Carcinoma Overview Mask,S146_HandE.ndpi [macro mask image],9846338
Lung Carcinoma Overview Mask,S147_HandE.ndpi [macro mask image],9846356
Lung Carcinoma Overview Mask,S148_HandE.ndpi [macro mask image],9846290
Lung Carcinoma Overview Mask,S149_HandE.ndpi [macro mask image],9846371
Lung Carcinoma Overview Mask,S14_HandE.ndpi [macro mask image],9846230
Lung Carcinoma Overview Mask,S150_HandE.ndpi [macro mask image],9846461
Lung Carcinoma Overview Mask,S151_HandE.ndpi [macro mask image],9846530
Lung Carcinoma Overview Mask,S152_HandE.ndpi [macro mask image],9846302
Lung Carcinoma Overview Mask,S153_HandE.ndpi [macro mask image],9846395
Lung Carcinoma Overview Mask,S154_HandE.ndpi [macro mask image],9846533
Lung Carcinoma Overview Mask,S156_HandE.ndpi [macro mask image],9846311
Lung Carcinoma Overview Mask,S157_HandE.ndpi [macro mask image],9846470
Lung Carcinoma Overview Mask,S158_HandE.ndpi [macro mask image],9846446
Lung Carcinoma Overview Mask,S159_HandE.ndpi [macro mask image],9846491
Lung Carcinoma Overview Mask,S160_HandE.ndpi [macro mask image],9846503
Lung Carcinoma Overview Mask,S161_HandE.ndpi [macro mask image],9846272
Lung Carcinoma Overview Mask,S162_HandE.ndpi [macro mask image],9846203
Lung Carcinoma Overview Mask,S163_HandE.ndpi [macro mask image],9846449
Lung Carcinoma Overview Mask,S164_HandE.ndpi [macro mask image],9846392
Lung Carcinoma Overview Mask,S165_HandE.ndpi [macro mask image],9846377
Lung Carcinoma Overview Mask,S166_HandE.ndpi [macro mask image],9846326
Lung Carcinoma Overview Mask,S167_HandE.ndpi [macro mask image],9846227
Lung Carcinoma Overview Mask,S168_HandE.ndpi [macro mask image],9846368
Lung Carcinoma Overview Mask,S169_HandE.ndpi [macro mask image],9846248
Lung Carcinoma Overview Mask,S16_HandE.ndpi [macro mask image],9846278
Lung Carcinoma Overview Mask,S170_HandE.ndpi [macro mask image],9846218
Lung Carcinoma Overview Mask,S17_HandE.ndpi [macro mask image],9846374
Lung Carcinoma Overview Mask,S18_HandE.ndpi [macro mask image],9846494
Lung Carcinoma Overview Mask,S19_HandE.ndpi [macro mask image],9846350
Lung Carcinoma Overview Mask,S1_HandE.ndpi [macro mask image],9846305
Lung Carcinoma Overview Mask,S20_HandE.ndpi [macro mask image],9846479
Lung Carcinoma Overview Mask,S21_HandE.ndpi [macro mask image],9846281
Lung Carcinoma Overview Mask,S22_HandE.ndpi [macro mask image],9846455
Lung Carcinoma Overview Mask,S23_HandE.ndpi [macro mask image],9846233
Lung Carcinoma Overview Mask,S24_HandE.ndpi [macro mask image],9846299
Lung Carcinoma Overview Mask,S25_HandE.ndpi [macro mask image],9846410
Lung Carcinoma Overview Mask,S26_HandE.ndpi [macro mask image],9846383
Lung Carcinoma Overview Mask,S27_HandE.ndpi [macro mask image],9846344
Lung Carcinoma Overview Mask,S28_HandE.ndpi [macro mask image],9846437
Lung Carcinoma Overview Mask,S29_HandE.ndpi [macro mask image],9846209
Lung Carcinoma Overview Mask,S2_HandE.ndpi [macro mask image],9846389
Lung Carcinoma Overview Mask,S32_HandE.ndpi [macro mask image],9846482
Lung Carcinoma Overview Mask,S36_HandE.ndpi [macro mask image],9846242
Lung Carcinoma Overview Mask,S38_HandE.ndpi [macro mask image],9846341
Lung Carcinoma Overview Mask,S39_HandE.ndpi [macro mask image],9846320
Lung Carcinoma Overview Mask,S3_HandE.ndpi [macro mask image],9846239
Lung Carcinoma Overview Mask,S40_HandE.ndpi [macro mask image],9846254
Lung Carcinoma Overview Mask,S41_HandE.ndpi [macro mask image],9846347
Lung Carcinoma Overview Mask,S42_HandE.ndpi [macro mask image],9846497
Lung Carcinoma Overview Mask,S43_HandE.ndpi [macro mask image],9846422
Lung Carcinoma Overview Mask,S46_HandE.ndpi [macro mask image],9846428
Lung Carcinoma Overview Mask,S47_HandE.ndpi [macro mask image],9846257
Lung Carcinoma Overview Mask,S4_HandE.ndpi [macro mask image],9846488
Lung Carcinoma Overview Mask,S50_HandE.ndpi [macro mask image],9846308
Lung Carcinoma Overview Mask,S51_HandE.ndpi [macro mask image],9846329
Lung Carcinoma Overview Mask,S53_HandE.ndpi [macro mask image],9846527
Lung Carcinoma Overview Mask,S55_HandE.ndpi [macro mask image],9846464
Lung Carcinoma Overview Mask,S56_HandE.ndpi [macro mask image],9846314
Lung Carcinoma Overview Mask,S57_HandE.ndpi [macro mask image],9846335
Lung Carcinoma Overview Mask,S58_HandE.ndpi [macro mask image],9846419
Lung Carcinoma Overview Mask,S59_HandE.ndpi [macro mask image],9846323
Lung Carcinoma Overview Mask,S60_HandE.ndpi [macro mask image],9846521
Lung Carcinoma Overview Mask,S61_HandE.ndpi [macro mask image],9846365
Lung Carcinoma Overview Mask,S62_HandE.ndpi [macro mask image],9846524
Lung Carcinoma Overview Mask,S63_HandE.ndpi [macro mask image],9846224
Lung Carcinoma Overview Mask,S65_HandE.ndpi [macro mask image],9846509
Lung Carcinoma Overview Mask,S66_HandE.ndpi [macro mask image],9846476
Lung Carcinoma Overview Mask,S67_HandE.ndpi [macro mask image],9846473
Lung Carcinoma Overview Mask,S68_HandE.ndpi [macro mask image],9846404
Lung Carcinoma Overview Mask,S69_HandE.ndpi [macro mask image],9846443
Lung Carcinoma Overview Mask,S6_HandE.ndpi [macro mask image],9846269
Lung Carcinoma Overview Mask,S71_HandE.ndpi [macro mask image],9846398
Lung Carcinoma Overview Mask,S72_HandE.ndpi [macro mask image],9846506
Lung Carcinoma Overview Mask,S73_HandE.ndpi [macro mask image],9846317
Lung Carcinoma Overview Mask,S74_HandE.ndpi [macro mask image],9846401
Lung Carcinoma Overview Mask,S76_HandE.ndpi [macro mask image],9846260
Lung Carcinoma Overview Mask,S77_HandE.ndpi [macro mask image],9846452
Lung Carcinoma Overview Mask,S78_HandE.ndpi [macro mask image],9846440
Lung Carcinoma Overview Mask,S79_HandE.ndpi [macro mask image],9846236
Lung Carcinoma Overview Mask,S7_HandE.ndpi [macro mask image],9846386
Lung Carcinoma Overview Mask,S80_HandE.ndpi [macro mask image],9846380
Lung Carcinoma Overview Mask,S81_HandE.ndpi [macro mask image],9846206
Lung Carcinoma Overview Mask,S82_HandE.ndpi [macro mask image],9846518
Lung Carcinoma Overview Mask,S83_HandE.ndpi [macro mask image],9846287
Lung Carcinoma Overview Mask,S84_HandE.ndpi [macro mask image],9846245
Lung Carcinoma Overview Mask,S85_HandE.ndpi [macro mask image],9846434
Lung Carcinoma Overview Mask,S86_HandE.ndpi [macro mask image],9846362
Lung Carcinoma Overview Mask,S87_HandE.ndpi [macro mask image],9846515
Lung Carcinoma Overview Mask,S88_HandE.ndpi [macro mask image],9846536
Lung Carcinoma Overview Mask,S89_HandE.ndpi [macro mask image],9846215
Lung Carcinoma Overview Mask,S8_HandE.ndpi [macro mask image],9846263
Lung Carcinoma Overview Mask,S90_HandE.ndpi [macro mask image],9846458
Lung Carcinoma Overview Mask,S91_HandE.ndpi [macro mask image],9846284
Lung Carcinoma Overview Mask,S92_HandE.ndpi [macro mask image],9846413
Lung Carcinoma Overview Mask,S93_HandE.ndpi [macro mask image],9846275
Lung Carcinoma Overview Mask,S94_HandE.ndpi [macro mask image],9846212
Lung Carcinoma Overview Mask,S96_HandE.ndpi [macro mask image],9846407
Lung Carcinoma Overview Mask,S97_HandE.ndpi [macro mask image],9846332
Lung Carcinoma Overview Mask,S9_HandE.ndpi [macro mask image],9846296
```

</details>

File: scripts/annotate/check_annotations.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+
+def printusage():
+    print('''
+Checks an IDR dataset to make sure each image/well has an annotation.
+Prints the images/wells which don't have any annotations to stdout.
+
+If an annotation.csv is provided the images will be checked if they have
+a corresponding entry in the annotation.csv. If it is omitted the images
+will simply be check if they have at least one annotation attached.
```
Comment (pwalczysko) [inline scripts/annotate/check_annotations.py:15]:
```suggestion
will simply be checked if they have at least one annotation attached.
```

Maybe it should say ``the images will simply pass the test if...`` ?

**Comment by pwalczysko** [comment]
The last two commits (help and the more precise checking for annotations) work as expected.

**Comment by pwalczysko** [comment]
Bug:
It does not seem to work on Screens, at least the one I tried.

```
(venv3) [pwalczysko@ome-idr-covid-1 annotate]$ python check_annotations.py Screen:2401
Traceback (most recent call last):
  File "check_annotations.py", line 109, in <module>
    for pl in screen.listChildren():
AttributeError: 'NoneType' object has no attribute 'listChildren'

```

File: scripts/annotate/check_annotations.py (hunk 1 above)
Comment (pwalczysko) [reply scripts/annotate/check_annotations.py:15]: I understand now that one is supposed to run the option **with annotations.csv** before one makes any annotation steps, to ease the workflow. Please add this to the help text.

**Comment by pwalczysko** [comment]
One RFE for the help text, otherwise good to go fmpov.

**Comment by pwalczysko** [comment]
Thank you.

**Review by pwalczysko — APPROVED** [review]

**Pull request merged**